### PR TITLE
chore: Handle async migrations & add migration tests

### PR DESCRIPTION
**Pull request by Cal-L**
## **Description**

This PR enables the ability to write both asynchronous and synchronous migrations. Moving forward, we should not have to worry about resolving the incoming states within migration themselves. This change wraps each migration in an asynchronous function, which resolves the state before sending it to the next migration. We also added unit tests for checking migration functionality using `createMigrate`

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

### **After**

The videos below shows that migrations are still functioning as expected. There is no change in user experience. In addition to E2E tests, the main QA test to check is testing upgrading from current release (7.18.0) to this branch.

Video of upgrade from 7.17.0 -> This branch. Example: Shows migration 35 being applied, which sets the displayed currency to ETH
https://github.com/MetaMask/metamask-mobile/assets/10508597/ab21c8f1-de46-40b4-85a7-5ece5fbe4425

Same flow going from 7.15.0 -> This branch
https://github.com/MetaMask/metamask-mobile/assets/10508597/4d03ddea-2391-4372-b553-f79c9bd5300f

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/store/migrations/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      }
+    }
+
+    expect(isPromiseMigrations).toEqual(true);
+  });
+});
+
+describe('migrations', () => {
+  beforeAll(() => {
+    process.env.NODE_ENV = 'production';
```
Comment (tommasini) [inline app/store/migrations/index.test.ts:72]: Curiousm why we need to have the node_env to prod here?

**Review by tommasini — DISMISSED** [review]
LGTM!!! This can be described with just this word -> AWESOME !!!

File: app/store/migrations/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import {
+  MigrationManifest,
+  PersistedState,
+  createMigrate,
+} from 'redux-persist';
+const defaultNodeEnv = process.env.NODE_ENV;
+jest.unmock('redux-persist');
+jest.mock('../../store', () => {});
+
+// Only test migrations 25 and up
```
Comment (tommasini) [inline app/store/migrations/index.test.ts:11]: If we are not testing the exceptions of sentry, this keeps to make sense? Should we test all the migrations?

Comment (Cal-L) [reply app/store/migrations/index.test.ts:11]: The earlier migrations do not capture exceptions and fails when accessing undefined states, which is why we start from 25 and up. This should include enough migrations for testing purposes

Comment (tommasini) [reply app/store/migrations/index.test.ts:11]: Ahh yes! because it doesn't capture it crashes because we are not mocking all the data! got it!

**Comment by chrisleewilcox** [comment]
@Cal-L can you take a look at the merge conflicts.

**Review by tommasini — APPROVED** [review]

**Review by cortisiko — APPROVED** [review]
This PR is QA passed.

**Pull request merged**